<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Compare varaince decomposition models on voom transformed RNA-seq data</title>

<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<style type="text/css">

/* padding for bootstrap navbar */
body {
  padding-top: 50px;
  padding-bottom: 40px;
}


/* offset scroll position for anchor links (for fixed navbar)  */
.section h2 {
  padding-top: 55px;
  margin-top: -55px;
}
.section h3 {
  padding-top: 55px;
  margin-top: -55px;
}



/* don't use link color in navbar */
.dropdown-menu>li>a {
  color: black;
}

/* some padding for disqus */
#disqus_thread {
  margin-top: 45px;
}

</style>

<link rel="stylesheet" href="libs/font-awesome-4.1.0/css/font-awesome.min.css"/>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="libs/highlight/textmate.css"
      type="text/css" />
<script src="libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>




</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="libs/navigation-1.0/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ashlar</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li><a href="index.html">Home</a></li>
        <li><a href="about.html">About</a></li>
        <li><a href="license.html">License</a></li>
        <li><a href="https://github.com/mengyin/ashlar-fash">GitHub</a></li>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">


<h1 class="title">Compare varaince decomposition models on voom transformed RNA-seq data</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#check-model-2">Check model 2</a></li>
<li><a href="#check-model-3">Check Model 3</a></li>
<li><a href="#reference">Reference</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2016-05-02</p>
<p><strong>Code version:</strong> 8cbccb9c71ea875854d5b72ea2506bd5cd265b4a</p>
<p>Previously we used fash to analyze the variance decomposition for a simple random effect model <span class="math display">\[Y_{i}=\mu+\beta_{c(i)}+e_i,\]</span> where <span class="math inline">\(\beta_{c(i)}\sim N(0,\sigma^2)\)</span> and <span class="math inline">\(e_i\sim N(0,s^2)\)</span>, and we can estimate the PVE (<span class="math inline">\(\sigma^2/(\sigma^2+s^2)\)</span>).</p>
<p>However, for RNA-seq data, after voom transformation we have a weighted linear regression model for each gene <span class="math inline">\(g\)</span> (suppose <span class="math inline">\(N\)</span> is number of samples in each group and <span class="math inline">\(K\)</span> is number of groups, so <span class="math inline">\(Y_g\)</span> is <span class="math inline">\(NK\)</span>-vector and <span class="math inline">\(\beta_g\)</span> is <span class="math inline">\(K\)</span>-vector): <span class="math display">\[Y_g=X\beta_g+W_g^{-1/2}e_g,\]</span> where <span class="math inline">\(X\)</span> is the design matrix, <span class="math inline">\(W_g\)</span> is the diagonal weight matrix (weights inversly proportional to <span class="math inline">\(Var(Y_g)\)</span>, and <span class="math inline">\(e_g\sim N(0,s_g^2I_{NK \times NK})\)</span>). Note that <span class="math inline">\(W_g\)</span> depends on count level, but <span class="math inline">\(s_g^2\)</span> doesn’t! <span class="math inline">\(s_g^2\)</span> is the de-trended error variance.</p>
<p>Now our goal is to model the random effect <span class="math inline">\(X\beta_g\)</span>. A few possible ways (note: <span class="math inline">\(\sigma^2\)</span> doesn’t depend on count level!):</p>
<ul>
<li><p>Model 1: <span class="math inline">\(Y_g=\mu_g+X\beta_g+W_g^{-1/2}e_g\)</span> and <span class="math inline">\(\beta_g\sim N(0,\sigma_g^2 I_{K \times K})\)</span>. However this model’s ANOVA F-statistic doesn’t have close form distribution!</p></li>
<li><p>Model 2: <span class="math inline">\(Y_g=\mu_g+X\beta_g+W_g^{-1/2}e_g\)</span> and <span class="math inline">\(\beta_g\sim N(0,\sigma_g^2N(X^TW_gX)^{-1})\)</span>. In the balanced design model, this assumption means that <span class="math inline">\(\beta_g\)</span>’s variance is inversely proportional to the average observation’s weights within each group. So the ANOVA’s F-statistic will follow the distribution (Johnson 2007) <span class="math display">\[F_g\sim (1+N\sigma_g^2/s_g^2)F(K-1,NK-K).\]</span></p></li>
<li><p>Model 3: <span class="math inline">\(Y_g=\mu_g+W_g^{-1/2}X\tilde{\beta_g}+W_g^{-1/2}e_g\)</span> and <span class="math inline">\(\tilde{\beta}_g\sim N(0,\sigma_g^2 I_{K \times K})\)</span>. Note that <span class="math inline">\(\tilde{\beta}\)</span> is different from <span class="math inline">\(\beta\)</span> defined in voom paper and the interpretation changes!</p></li>
</ul>
<p>We check whether Model 2 and Model 3’s assumption is sensible on GTEx data.</p>
<p>First define functions to generate RNA-Seq datasets (we use GTEx data). Here we use data from multiple tissues (conditions), and analyze the conditional effects using F-test, limma or fash (on voom-transformed data).</p>
<pre class="r"><code>library(edgeR)</code></pre>
<pre><code>Loading required package: limma</code></pre>
<pre class="r"><code>library(limma)
library(qvalue)</code></pre>
<pre><code>Warning: replacing previous import by &#39;grid::arrow&#39; when loading &#39;qvalue&#39;</code></pre>
<pre><code>Warning: replacing previous import by &#39;grid::unit&#39; when loading &#39;qvalue&#39;</code></pre>
<pre class="r"><code>source(&quot;../code/fash.R&quot;)</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: evd</code></pre>
<pre><code>
Attaching package: &#39;evd&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:ashr&#39;:

    dens</code></pre>
<pre class="r"><code>datamaker = function(args){  
  if (length(args$tissue)&lt;args$ngroup){
    stop(&quot;Not enough tissues!&quot;)
  }
  
  rawdata = list()
  for (i in 1:args$ngroup){
    rawdata[[i]] = read.table(paste0(&quot;/Volumes/PERSONAL/MS/gtex/tissues/&quot;,
                                     args$tissue[i],&quot;.txt&quot;),header=TRUE)
  }
  
  # randomly choose samples for each gene
  counts = NULL
  for (i in 1:args$ngroup){
    counts = cbind(counts, selectsample(rawdata[[i]], args$nsamp, args$breaksample))
  }
  rm(&quot;rawdata&quot;)
  counts = shufflegroup(counts, args$ngroup, args$nsamp, args$breaksample)

  # Remove genes without any reads
  counts = counts[apply(counts,1,sum)&gt;0,]
  
  # Take the top Ngene high-expressed genes
  counts = counts[sort(order(rowSums(counts),decreasing=TRUE)[1:args$ngene]),]
  
  condition = factor(rep(1:args$ngroup,each=args$nsamp))
  
  # Voom transformation + limma
  voomlim = voom_transform(counts, condition)
  
  # f-test
  ftest = varcomp_voom(voomlim$v, condition)
  
  # meta data
  meta = list(args=args)
  
  # input data
  input = list(counts=counts, condition=condition, 
               v=voomlim$v, avelogcpm=voomlim$avelogcpm, 
               fstats.limma=voomlim$fstats, pve.limma=voomlim$pve, pvalue.limma=voomlim$pvalue,
               df1.limma=args$ngroup-1, df2.limma=voomlim$lim$df.total[1],
               fstats.ftest=ftest$fstats, pve.ftest=ftest$pve)  
  data = list(meta=meta,input=input)
  return(data)
}

# Voom transformation
voom_transform = function(counts, condition, W=NULL){
  dgecounts = calcNormFactors(DGEList(counts=counts,group=condition))
  
  if (is.null(W)){
    design = model.matrix(~condition)
  }else{
    design = model.matrix(~condition+W)
  }
  
  v = voom(dgecounts,design,plot=FALSE)
  lim = lmFit(v)
  lim = eBayes(lim[,-1])
  pve = pve_ftest(length(condition)/length(unique(condition)), 
                  lim$s2.post*lim$F, lim$s2.post)
  
  dgecounts = estimateDisp(dgecounts, design, robust=TRUE)
  avelogcpm = dgecounts$AveLogCPM
  
  return(list(v=v, lim=lim, avelogcpm=avelogcpm, 
              fstats=lim$F, pvalue=lim$F.p.value, pve=pve))
}

varcomp_voom = function(v, condition){
  combo = cbind(v$E,v$weights)
  tmp = apply(combo,1,voom_ftest,condition=condition)
  MST=tmp[2,]
  MSE=tmp[3,]
  pve = pve_ftest(length(condition)/length(unique(condition)), MST, MSE)
  return(list(fstats=tmp[1,], MST=MST, MSE=MSE, pve=pve))
}

voom_ftest = function(ynweights,condition){
  y = ynweights[1:(length(ynweights)/2)]
  weights = ynweights[(length(ynweights)/2+1):length(ynweights)]
  fit = lm(y~condition, weights=weights)
  MST = anova(fit)$M[1]
  MSE = anova(fit)$M[length(anova(fit)$M)]
  fstat = anova(fit)$F[1]
  return(c(fstat,MST,MSE))
}

# randomly subsample data for each gene
# gene: a vector of reads for one gene
# nsamp: # of samples wanted
sampleingene = function(gene, nsamp){
  sample = sample(length(gene),nsamp)
  return(c(gene[sample]))
}

# Randomly select samples
# counts: full count matrix
# nsamp: # of samples wanted
# breaksample: flag, if select different samples for each gene
selectsample = function(counts, nsamp, breaksample){
  if (breaksample==FALSE){
    subsample = sample(1:dim(counts)[2],nsamp)
    counts = counts[,subsample]
  }else{
    counts = t(apply(counts, 1, sampleingene, nsamp=nsamp))
  }
  return(counts)
}

# randomly shuffle the groups&#39; order for each gene
# to avoid the unequal library size issue
shufflegroup = function(counts, ngroup, nsamp, breaksample){
  if (breaksample==TRUE){
    return(t(apply(counts, 1, shufflegroup_vec, ngroup=ngroup, nsamp=nsamp)))
  }else{
    return(counts)
  }
}

shufflegroup_vec = function(gene, ngroup, nsamp){
  idx = sample(ngroup,replace=FALSE)
  idx = rep((idx-1)*nsamp,each=nsamp)+1:nsamp
  return(gene[idx])
}

pve_ftest = function(nsamp,MST,MSE){
  sigma.c2.hat = pmax(0,(MST-MSE)/nsamp)
  sigma.e2.hat = MSE
  return(sigma.c2.hat/(sigma.c2.hat+sigma.e2.hat))
}

pve_plugin = function(nsamp,fashobj){
  (exp(fashobj$PosteriorMean.logf)-1)/(exp(fashobj$PosteriorMean.logf)-1+nsamp)
}</code></pre>
<p>Sample dataset (independent genes): for each gene, randomly choose 50 samples from each of the 10 tissues. We randomly shuffle the order of the 10 tissues for each gene, such that the library sizes of all samples are similar so we won’t have the unequal library sizes issue.</p>
<pre class="r"><code>set.seed(1234)
# Generate a pure null dataset
args = list(tissue=c(&quot;Brain-Amygdala&quot;,&quot;Brain-Caudate(basalganglia)&quot;,&quot;Brain-CerebellarHemisphere&quot;,
                     &quot;Brain-Cerebellum&quot;,&quot;Brain-Cortex&quot;,&quot;Brain-FrontalCortex(BA9)&quot;,
                     &quot;Brain-Nucleusaccumbens(basalganglia)&quot;,&quot;Brain-Hippocampus&quot;,&quot;Brain-Hypothalamus&quot;,
                     &quot;Brain-Substantianigra&quot;), 
            ngene=10000, ngroup=10, nsamp=50, breaksample=TRUE)
test = datamaker(args)</code></pre>
<div id="check-model-2" class="section level2">
<h2>Check model 2</h2>
<p>Since Model 2 assumes that <span class="math inline">\(Var(\beta)=\sigma^2N(X^TWX)^{-1}\)</span>, where <span class="math inline">\(\sigma^2\)</span>’s are exchangeable across genes and do not depend on count levels. Therefore the variances of the normalized effect sizes <span class="math inline">\(\beta^*_g:=\beta_g/\sqrt{N(X^TWX)^{-1}}\)</span> should be exchangeable. We see that there are no obvious trend between the distribution of <span class="math inline">\(sd(\hat{\beta^*})\)</span> and the average log-cpm values if we remove the leverage points on the right.</p>
<pre class="r"><code># compute the normalizing factors (X&#39;WX) for each gene
scalew_vec = function(weights, condition){
  return(t(apply(weights,1,scalew, model.matrix(~condition)[,-1])))
}

# X: design matrix, W: weights
# return new weights: diagonal of (X&#39;WX)^(-1)
scalew = function(W, X){
  diag(solve(t(X)%*%diag(W)%*%X))
}

# fit the regression coefficients beta
lim = lmFit(test$input$v,model.matrix(~test$input$condition))

# normalize beta by sqrt(X&#39;WX)^(-1)
weights_beta = scalew_vec(test$input$v$weights,test$input$condition)
coef_w = lim$coefficients[,-1]/sqrt(test$meta$args$nsamp*weights_beta) # no intercept

# plot sd&#39;s of normalized beta against average log-cpm
sd_coef_w = apply(coef_w, 1, sd)
scatter.smooth(test$input$avelogcpm, sd_coef_w,col=2,ylim=c(0,10),
               xlab=&quot;average log-cpm&quot;, ylab=&quot;sd(beta_star)&quot;,main=&quot;Model 2&quot;)</code></pre>
<p><img src="figure/alt_voom_ftest.Rmd/unnamed-chunk-3-1.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># remove some leverage points on the right
scatter.smooth(test$input$avelogcpm[test$input$avelogcpm&lt;8], sd_coef_w[test$input$avelogcpm&lt;8],
               ylim=c(0,10),col=2,xlab=&quot;average log-cpm&quot;, ylab=&quot;sd(beta_star)&quot;,main=&quot;Model 2&quot;)</code></pre>
<p><img src="figure/alt_voom_ftest.Rmd/unnamed-chunk-3-2.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="check-model-3" class="section level2">
<h2>Check Model 3</h2>
<p>We fit <span class="math inline">\(\tilde{\beta}\)</span> by running a simple linear regression data on centered and normalized data <span class="math inline">\(\tilde{Y}_g:=W_g^{1/2}(Y_g-\mu_g)=X\tilde{\beta}_g+e_g\)</span>. Since Model 3 assumes that <span class="math inline">\(\tilde{\beta}_g\sim N(0,\sigma_g^2)\)</span> and <span class="math inline">\(\sigma_g^2\)</span> does not depend on count level, we can plot <span class="math inline">\(sd(\hat{\tilde{\beta}})\)</span> against the average log-cpm values and check if there is obvious trend. Similar to the above discussion for Model 2, there is no obvious relationship between them if we drop some leverage points on the right.</p>
<pre class="r"><code># Y_tilde is centered and normalized by error weights
Y_tilde = (test$input$v$E-lim$coefficients[,1])*sqrt(test$input$v$weights)

# fit beta_tilde
reg = function(y,condition){
  m = lm(y~condition)
  m$coef+c(0,rep(m$coef[1],length(m$coef)-1))
}
coef_tilde = t(apply(Y_tilde,1,reg, condition=test$input$condition))

# plot plot sd&#39;s of beta_tilde against average log-cpm
sd_coef_tilde = apply(coef_tilde[,-1],1,sd)
scatter.smooth(test$input$avelogcpm, sd_coef_tilde,col=2,ylim=c(0,10),
               xlab=&quot;average log-cpm&quot;, ylab=&quot;sd(beta_tilde)&quot;,main=&quot;Model 3&quot;)</code></pre>
<p><img src="figure/alt_voom_ftest.Rmd/unnamed-chunk-4-1.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># remove some leverage points on the right
scatter.smooth(test$input$avelogcpm[test$input$avelogcpm&lt;8], sd_coef_tilde[test$input$avelogcpm&lt;8],
               col=2,ylim=c(0,10),xlab=&quot;average log-cpm&quot;, ylab=&quot;sd(beta_tilde)&quot;,main=&quot;Model 3&quot;)</code></pre>
<p><img src="figure/alt_voom_ftest.Rmd/unnamed-chunk-4-2.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
<p>Actually in this dataset, Model 2’s <span class="math inline">\(sd(\hat{\tilde{\beta}})\)</span> and Model 3’s <span class="math inline">\(sd(\hat{\tilde{\beta}})\)</span> are almost same.</p>
<pre class="r"><code>plot(sd_coef_w,sd_coef_tilde, xlab=&quot;Model 2: sd(beta_star)&quot;,ylab=&quot;Model 3: sd(beta_tilde)&quot;)
abline(0,1,col=2)</code></pre>
<p><img src="figure/alt_voom_ftest.Rmd/unnamed-chunk-5-1.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="reference" class="section level2">
<h2>Reference</h2>
<p>[1] Johnson, Valen E. “Properties of Bayes factors based on test statistics.” Scandinavian Journal of Statistics 35.2 (2008): 354-368. [2] Johnson, Valen E. “Bayes factors based on test statistics.” Journal of the Royal Statistical Society: Series B (Statistical Methodology) 67.5 (2005): 689-701.</p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.2.3 (2015-12-10)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.10.5 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
[1] truncdist_1.0-1 evd_2.3-2       ashr_1.0.12     qvalue_2.2.0   
[5] edgeR_3.10.5    limma_3.24.15   knitr_1.12.3   

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.4       magrittr_1.5      MASS_7.3-45      
 [4] splines_3.2.3     statmod_1.4.24    pscl_1.4.9       
 [7] munsell_0.4.3     doParallel_1.0.10 lattice_0.20-33  
[10] SQUAREM_2014.8-1  colorspace_1.2-6  foreach_1.4.3    
[13] stringr_1.0.0     plyr_1.8.3        tools_3.2.3      
[16] parallel_3.2.3    grid_3.2.3        gtable_0.1.2     
[19] iterators_1.0.8   htmltools_0.3     yaml_2.1.13      
[22] digest_0.6.9      reshape2_1.4.1    ggplot2_2.0.0    
[25] formatR_1.2.1     codetools_0.2-14  evaluate_0.8     
[28] rmarkdown_0.9.5   stringi_1.0-1     scales_0.3.0     
[31] locfit_1.5-9.1    truncnorm_1.0-7  </code></pre>
</div>


<!-- some extra javascript for older browsers -->
<script type="text/javascript" src="libs/polyfill.js"></script>

<script>

// manage active state of menu based on current page
$(document).ready(function () {

    // active menu
    href = window.location.pathname
    href = href.substr(href.lastIndexOf('/') + 1)
    $('a[href="' + href + '"]').parent().addClass('active');

    // manage active menu header
    if (href.startsWith('authoring_'))
      $('a[href="' + 'authoring' + '"]').parent().addClass('active');
    else if (href.endsWith('_format.html'))
      $('a[href="' + 'formats' + '"]').parent().addClass('active');
    else if (href.startsWith('developer_'))
      $('a[href="' + 'developer' + '"]').parent().addClass('active');

});

</script>



</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
