<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Mengyin Lu" />

<meta name="date" content="2016-01-20" />

<title>Compare RNA-seq gene-wise variance decomposition via glm vs voom using GTEx data</title>

<script src="libs/jquery-1.11.0/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.1/css/united.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.1/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/respond.min.js"></script>

<style type="text/css">

/* padding for bootstrap navbar */
body {
  padding-top: 50px;
  padding-bottom: 40px;
}


/* offset scroll position for anchor links (for fixed navbar)  */
.section h2 {
  padding-top: 55px;
  margin-top: -55px;
}
.section h3 {
  padding-top: 55px;
  margin-top: -55px;
}



/* don't use link color in navbar */
.dropdown-menu>li>a {
  color: black;
}

/* some padding for disqus */
#disqus_thread {
  margin-top: 45px;
}

</style>

<link rel="stylesheet" href="libs/font-awesome-4.1.0/css/font-awesome.min.css"/>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="libs/highlight/textmate.css"
      type="text/css" />
<script src="libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">


<div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ashlar</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li><a href="index.html">Home</a></li>
        <li><a href="about.html">About</a></li>
        <li><a href="license.html">License</a></li>
        <li><a href="https://github.com/jhsiao999/ashlar">GitHub</a></li>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">
<h1 class="title">Compare RNA-seq gene-wise variance decomposition via glm vs voom using GTEx data</h1>
<h4 class="author"><em>Mengyin Lu</em></h4>
<h4 class="date"><em>2016-01-20</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#objectives">Objectives</a></li>
<li><a href="#simulation">Simulation</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2016-02-10</p>
<p><strong>Code version:</strong> 4805492d8e6eeaa8da3fd2ee38cdca1eb6c816ea</p>
<p>This is a template for writing reports with R Markdown.</p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<p>Suppose we have RNA-seq data of <span class="math inline">\(N\)</span> samples from <span class="math inline">\(K\)</span> groups. We want to estimate <span class="math inline">\(r_g\)</span>, the ratio of within-group variation and between-group variation for each gene, i.e.Â do gene-wise variance decompostion analysis.</p>
<p>Two possible ways:</p>
<ol style="list-style-type: decimal">
<li>Suppose the model of counts is:
\begin{align}
Y_{gi} &amp;\sim Pois(\lambda_{gi}), \\
\log(\lambda_{gi}) &amp;= \mu_g+\beta_{g,c(i)}+e_{gi}, \\
e_{gi}\sim N(0,\sigma_{ge}^2), &amp;\quad \beta_{g,c(i)}\sim N(0,\sigma_{gc}^2).
\end{align}
<p>Then <span class="math inline">\(\hat{r}_g^{(GLM)} = \frac{\hat{\sigma_{ge}^2}}{\hat{\sigma_{gc}^2}}\)</span>.</p></li>
<li><p>First apply voom transformation on counts, then do ANOVA for the transformed data: <span class="math inline">\(\hat{r}_g^{(voom)} = MST_g/MSE_g\)</span>.</p></li>
</ol>
</div>
<div id="simulation" class="section level2">
<h2>Simulation</h2>
<p>Extract GTEx RNA-seq data from 10 tissues, with 50 samples for each tissue.</p>
<pre class="r"><code>library(edgeR)</code></pre>
<pre><code>Loading required package: limma</code></pre>
<pre class="r"><code>library(limma)

datamaker = function(args){
  counts = NULL
  for (tis in args$tissue){
    rawdata = read.table(paste0(&quot;/Volumes/PERSONAL/MS/gtex/tissues/&quot;,tis,&quot;.txt&quot;),header=TRUE)
    counts = cbind(counts,as.matrix(rawdata[,1:args$Nsamp]))
  }
  rm(&quot;rawdata&quot;)  

  # Remove genes without any reads
  counts = counts[apply(counts,1,sum)&gt;0,]

  condition = factor(rep(1:length(args$tissue),each=args$Nsamp))
  
  # Voom transformation
  dgecounts = DGEList(counts=counts,group=condition)
  dgecounts = calcNormFactors(dgecounts)
  design = model.matrix(~condition)
  v = voom(dgecounts,design,plot=FALSE)
  dgecounts = estimateDisp(dgecounts, design, robust=TRUE)
  avelogcpm = dgecounts$AveLogCPM
  
  # meta data
  meta = list(args=args)
  
  # input data
  input = list(counts=counts, condition=condition, v=v, avelogcpm=avelogcpm)  
  data = list(meta=meta,input=input)
  return(data)
}

# Compute the ratios (sigma_c/sigma_e) for each gene
# Model: y~Pois(lambda), lambda=mu+b_cond+e, where b_cond~N(0,sigma_c^2), e~N(0,sigma_e^2)
varcomp_glmm = function(counts, condition){
  apply(counts,1,glmm_ratio,condition=condition)
}

library(lme4)</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre class="r"><code>glmm_ratio = function(y,condition){
  obs = factor(1:length(y))
  fit = try(glmer(y~(1|condition)+(1|obs),family=poisson))
  
  # Ratio = sigma_c/sigma_e
  if (class(fit)==&quot;try-error&quot;){
    ratio = NA
  }else{
    ratio = sqrt(as.numeric(summary(fit)$varcor$condition)/
                   as.numeric(summary(fit)$varcor$obs))
  }
  
  return(ratio)
}

varcomp_voom = function(v, condition){
  combo = cbind(v$E,v$weights)
  apply(combo,1,voom_ratio,condition=condition)
}

voom_ratio = function(ynweights,condition){
  y = ynweights[1:(length(ynweights)/2)]
  weights = ynweights[(length(ynweights)/2+1):length(ynweights)]
  fit = lm(y~condition, weights=weights)
  ratio = sqrt(max(anova(fit)$F[1]-1,0)/length(unique(condition)))
  return(ratio)
}</code></pre>
<pre class="r"><code>args = list(tissue=c(&quot;Brain-Amygdala&quot;,
                     &quot;Brain-Anteriorcingulatecortex(BA24)&quot;,
                     &quot;Brain-Caudate(basalganglia)&quot;,
                     &quot;Brain-CerebellarHemisphere&quot;,
                     &quot;Brain-Cerebellum&quot;,
                     &quot;Brain-Cortex&quot;,
                     &quot;Brain-FrontalCortex(BA9)&quot;,
                     &quot;Brain-Hippocampus&quot;,
                     &quot;Brain-Hypothalamus&quot;,
                     &quot;Brain-Nucleusaccumbens(basalganglia)&quot;),
            Nsamp=50)
#data = datamaker(args)</code></pre>
<p>Compute <span class="math inline">\(\hat{r}_g^{(GLM)}\)</span> and <span class="math inline">\(\hat{r}_g^{(voom)}\)</span> for top 10,000 genes and compare them.</p>
<pre class="r"><code>load(&quot;/Volumes/PERSONAL/MS/fash/appli_gtex.RData&quot;)
#r.voom = varcomp_voom(data$input$v[1:10000,], data$input$condition)
#r.glmm = varcomp_glmm(data$input$counts[1:10000,], data$input$condition)

plot(log(r.voom), log(r.glmm), xlab=&quot;ratio_voom&quot;, ylab=&quot;ratio_glmm&quot;)</code></pre>
<p><img src="figure/vd_glm_vs_voom.Rmd/unnamed-chunk-3-1.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(data$input$avelogcpm[1:10000],log(r.voom),xlab=&quot;Ave log-cpm&quot;,ylab=&quot;ratio&quot;,main=&quot;Ratio=sigma_c/sigma_e (Voom)&quot;)</code></pre>
<p><img src="figure/vd_glm_vs_voom.Rmd/unnamed-chunk-3-2.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>plot(data$input$avelogcpm[1:10000],log(r.glmm),xlab=&quot;Ave log-cpm&quot;,ylab=&quot;ratio&quot;,main=&quot;Ratio=sigma_c/sigma_e (GLMM)&quot;)</code></pre>
<p><img src="figure/vd_glm_vs_voom.Rmd/unnamed-chunk-3-3.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
<p>Seems <span class="math inline">\(\hat{r}_g^{(GLM)}\)</span> can be very unstable when log-cpm is low? <span class="math inline">\(\hat{r}_g^{(voom)}\)</span> is a smooth function of log-cpm?</p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.2.3 (2015-12-10)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.10.5 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lme4_1.1-10   Matrix_1.2-3  edgeR_3.10.5  limma_3.24.15 knitr_1.12   

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.2     lattice_0.20-33 digest_0.6.8    MASS_7.3-45    
 [5] grid_3.2.3      nlme_3.1-122    formatR_1.2.1   magrittr_1.5   
 [9] evaluate_0.8    stringi_1.0-1   minqa_1.2.4     nloptr_1.0.4   
[13] rmarkdown_0.9.2 splines_3.2.3   tools_3.2.3     stringr_1.0.0  
[17] yaml_2.1.13     htmltools_0.3  </code></pre>
</div>


<!-- some extra javascript for older browsers -->
<script type="text/javascript" src="libs/polyfill.js"></script>

<script>

// manage active state of menu based on current page
$(document).ready(function () {

    // active menu
    href = window.location.pathname
    href = href.substr(href.lastIndexOf('/') + 1)
    $('a[href="' + href + '"]').parent().addClass('active');

    // manage active menu header
    if (href.startsWith('authoring_'))
      $('a[href="' + 'authoring' + '"]').parent().addClass('active');
    else if (href.endsWith('_format.html'))
      $('a[href="' + 'formats' + '"]').parent().addClass('active');
    else if (href.startsWith('developer_'))
      $('a[href="' + 'developer' + '"]').parent().addClass('active');

});

</script>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
